corehq/blobs/metadata.py

```diff
@@ -0,0 +1,89 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from collections import defaultdict
+from datetime import datetime, timedelta
+from itertools import chain
+
+from corehq.sql_db.routers import get_cursor
+from corehq.sql_db.util import split_list_by_db_partition
+from corehq.util.datadog.gauges import datadog_counter
+
+from .models import BlobMeta
+
+
+class MetaDB(object):
+    """Blob metadata database interface
+
+    This class manages persistence of blob metadata in a SQL database.
+    """
+
+    def new(self, **blob_meta_args):
+        """Get a new `BlobMeta` object
+
+        :param **blob_meta_args: See `AbstractBlobDB.put` for argument
+        documentation. If the given `meta` argument is a `BlobMeta`
+        object it will be returned instead of creating a new instance.
+        :raises: `ValueError` if the given `meta` object is a `BlobMeta`
+        instance that has already been saved.
+        """
+        if "meta" in blob_meta_args:
+            if len(blob_meta_args) > 1:
+                raise ValueError(
+                    "keyword arguments are incompatible with `meta` argument")
+            return blob_meta_args["meta"]
+        timeout = blob_meta_args.pop("timeout", None)
+        meta = BlobMeta(**blob_meta_args)
+        if not meta.domain:
+            raise TypeError("domain is required")
+        if not meta.parent_id:
+            raise TypeError("parent_id is required")
+        if meta.type_code is None:
+            raise TypeError("type_code is required")
+        if timeout is not None:
+            if "expires_on" in blob_meta_args:
+                raise ValueError("pass one: timeout or expires_on")
+            meta.expires_on = _utcnow() + timedelta(minutes=timeout)
+        return meta
+
+    def put(self, meta):
+        """Save `BlobMeta` in the metadata database"""
+        meta.save()
+        length = meta.content_length
+        datadog_counter('commcare.blobs.added.count')
+        datadog_counter('commcare.blobs.added.bytes', value=length)
+        if meta.expires_on is not None:
+            datadog_counter('commcare.temp_blobs.count')
+            datadog_counter('commcare.temp_blobs.bytes_added', value=length)
+
+    def delete(self, key, content_length):
```

Comment:
why not use the `meta` object here like in `bulk_delete`? That way you could delete it directly instead of having to delete by `key`

Reply:
The blob db `delete(key)` method does not take a meta object, and it would require an extra SQL query to load it.

Reply:
one query to load + 1 delete vs N delete queries (one to each DB)